Rename `ConstPropLint` to `KnownPanicsLint`

gurry opened the pull request:
`OverflowLint` is a clearer name because it communicates what the lint does instead of the underlying mechanism it uses (const propagation) which should be of secondary concern. 

`OverflowLint` isn't the most accurate name because the lint looks for other errors as well such as division by zero not just overflows, but I couldn't think of another equally succinct name.

As a part of this change. I've also added/updated some of the comments.

cc @RalfJung @oli-obk for visibility in case you go looking for the lint using the old name.

Edit:

Changed the name from `OverflowLint` to `KnownPanicsLint`

gurry commented on compiler/rustc_mir_transform/src/overflow_lint.rs:
```diff
@@ -37,31 +40,28 @@ impl<'tcx> MirLint<'tcx> for ConstPropLint {
-        trace!("ConstPropLint starting for {:?}", def_id);
+        trace!("OverflowLint starting for {:?}", def_id);
 
-        // FIXME(oli-obk, eddyb) Optimize locals (or even local paths) to hold
```
emoved this comment as it talks about "optimizing" locals while the code does not optimize anything any more. The comment is a relict from the glorious past when it used to be a full-blooded const propagator.

gurry commented:
Adding one more to the mix, `ConstOpsLint` since it checks that _operations_ on _constants_ do not panic.

Let's hear from others. I'll change it to whatever we all agree on.

gurry commented:
How about just `KnownPanic` instead of `KnownRuntimePanic` for brevity?

gurry commented:
Or just `PanicsLint` or even `RuntimePanicsLint` 'cause "Known" does not seem to add much value.

gurry commented:
Or `PanickingOpsLint` (sorry I'm getting a bit carried away here).

Nilstrieb commented:
I think it adds a lot of value. `PanicsLint` just implies a lint that finds potential panics in the program, but `KnownPanicsLint` makes it clear that these are places we _know_ will panic.

gurry commented:
Makes sense. Let's go with `KnownPanicsLint`. 

Will wait for a day. If no one objects, will make the change tomorrow.

gurry committed:
Rename `ConstPropLint` to `KnownPanicsLint`

It is a clearer name because it communicates what the lint does
instead of the underlying mechanism it uses (const propagation).

gurry commented:
Renamed to `KnownPanicsLint` and ready for review

oli-obk commented:
r? @oli-obk 

@bors r+ rollup
